In [1]:
%load_ext autoreload
%autoreload 2

In [4]:
import pandas as pd
import numpy as np
from rule_scoring.rule_scorer import RuleScorer
import cProfile, pstats

In [5]:
om = pd.read_pickle('~/Downloads/omnyex_processed_for_profiling.pkl')
X = om.drop('chargeback', axis=1)
y = om['chargeback']

In [6]:
X = X.astype(float)

In [12]:
X_rules = (X >= 1).astype(int)

In [22]:
from rule_scoring.rule_scoring_methods import PerformanceScorer, LogRegScorer, RandomForestScorer
from rule_scoring.rule_score_scalers import ConstantScaler, MinMaxScaler
from rule_optimisation.optimisation_functions import Precision

In [15]:
X_rules.shape

(1780373, 50)

# PerformanceScorer with ConstantScaler

## Old

In [29]:
precision = Precision()

In [31]:
# rs = RuleScorer(scoring_class=PerformanceScorer(performance_func=precision.fit), scaling_class=ConstantScaler(limit=-100))
# rs.fit(X_rules, y)

In [34]:
filename = 'PerformanceScorer_ConstantScaler_old.dat'
rs = RuleScorer(scoring_class=PerformanceScorer(performance_func=precision.fit), scaling_class=ConstantScaler(limit=-100))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

In [35]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

Mon Jan 11 12:02:08 2021    PerformanceScorer_ConstantScaler_old.dat.dat

         128445 function calls (126874 primitive calls) in 0.737 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.737    0.737 {built-in method builtins.exec}
        1    0.000    0.000    0.737    0.737 <string>:1(<module>)
        1    0.000    0.000    0.737    0.737 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scorer.py:20(fit)
        1    0.000    0.000    0.736    0.736 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scoring_methods.py:21(fit)
        1    0.000    0.000    0.735    0.735 <__array_function__ internals>:2(apply_along_axis)
    656/1    0.003    0.000    0.735    0.735 {built-in method numpy.core._multiarray_umath.implement_array_function}
        1    0.001    0.001    0.735    0.735 /Users/jlaidler/venvs/argo/lib/python3.

~.7sec runtime - pretty quick already

# PerformanceScorer with MinMaxScaler

## Old

In [37]:
rs = RuleScorer(scoring_class=PerformanceScorer(performance_func=precision.fit), scaling_class=MinMaxScaler(min_value=-100, max_value=0))
rs.fit(X_rules, y)

In [38]:
filename = 'PerformanceScorer_MinMaxScalar_old.dat'
rs = RuleScorer(scoring_class=PerformanceScorer(performance_func=precision.fit), scaling_class=MinMaxScaler(min_value=-100, max_value=0))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

In [39]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

Mon Jan 11 12:02:25 2021    PerformanceScorer_MinMaxScalar_old.dat.dat

         128973 function calls (127387 primitive calls) in 0.730 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.730    0.730 {built-in method builtins.exec}
        1    0.000    0.000    0.730    0.730 <string>:1(<module>)
        1    0.000    0.000    0.730    0.730 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scorer.py:20(fit)
        1    0.000    0.000    0.728    0.728 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scoring_methods.py:21(fit)
    662/7    0.003    0.000    0.728    0.104 {built-in method numpy.core._multiarray_umath.implement_array_function}
        1    0.000    0.000    0.727    0.727 <__array_function__ internals>:2(apply_along_axis)
        1    0.001    0.001    0.727    0.727 /Users/jlaidler/venvs/argo/lib/python3.7/

~.7sec runtime - pretty quick already

# LogRegScorer with ConstantScaler

## Old

In [41]:
filename = 'LogRegScorer_ConstantScaler_old.dat'
rs = RuleScorer(scoring_class=LogRegScorer(), scaling_class=ConstantScaler(limit=-100))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

/Users/jlaidler/venvs/argo/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [42]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

Mon Jan 11 12:07:57 2021    LogRegScorer_ConstantScaler_old.dat.dat

         23026 function calls (22801 primitive calls) in 34.061 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   34.061   34.061 {built-in method builtins.exec}
        1    0.000    0.000   34.061   34.061 <string>:1(<module>)
        1    0.000    0.000   34.061   34.061 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scorer.py:20(fit)
        1    0.052    0.052   34.059   34.059 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scoring_methods.py:54(fit)
        1    0.000    0.000   34.007   34.007 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:1275(fit)
        1    0.000    0.000   32.389   32.389 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/joblib/parallel.py:946(__call__)
        2    0.000    0.000

~34 secs is mostly due to Sklearn's logreg fit method. Can't really improve that.

# LogRegScorer with MinMaxScaler

## Old

In [52]:
filename = 'LogRegScorer_MinMaxScaler_old.dat'
rs = RuleScorer(scoring_class=LogRegScorer(), scaling_class=MinMaxScaler(min_value=-100, max_value=0))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

/Users/jlaidler/venvs/argo/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [53]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

Mon Jan 11 15:15:41 2021    LogRegScorer_MinMaxScaler_old.dat.dat

         23498 function calls (23284 primitive calls) in 33.001 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   33.001   33.001 {built-in method builtins.exec}
        1    0.000    0.000   33.001   33.001 <string>:1(<module>)
        1    0.000    0.000   33.001   33.001 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scorer.py:20(fit)
        1    0.069    0.069   32.999   32.999 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scoring_methods.py:56(fit)
        1    0.000    0.000   32.929   32.929 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:1275(fit)
        1    0.000    0.000   30.946   30.946 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/joblib/parallel.py:946(__call__)
        2    0.000    0.000  

~34 secs is mostly due to Sklearn's logreg fit method. Can't really improve that.

# RandomForestScorer with ConstantScaler

## Old

In [54]:
filename = 'RandomForestScorer_ConstantScaler_old.dat'
rs = RuleScorer(scoring_class=RandomForestScorer(), scaling_class=ConstantScaler(limit=-100))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

In [55]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

Mon Jan 11 15:19:12 2021    RandomForestScorer_ConstantScaler_old.dat.dat

         222189 function calls (218999 primitive calls) in 204.249 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000  204.249  204.249 {built-in method builtins.exec}
        1    0.000    0.000  204.249  204.249 <string>:1(<module>)
        1    0.017    0.017  204.249  204.249 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scorer.py:20(fit)
        1    0.031    0.031  204.231  204.231 /Users/jlaidler/Documents/tigress/tigress/argo/argo/rule_scoring/rule_scoring/rule_scoring_methods.py:91(fit)
        1    0.001    0.001  204.183  204.183 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/sklearn/ensemble/_forest.py:272(fit)
        2    0.001    0.000  203.624  101.812 /Users/jlaidler/venvs/argo/lib/python3.7/site-packages/joblib/parallel.py:946(__call__)
      202    0.004    0.0

~204 secs is mostly due to Sklearn's rf fit method. Can't really improve that - use parameters to reduce time instead.

# RandomForestScorer with MinMaxScaler

## Old

In [ ]:
filename = 'RandomForestScorer_MinMaxScaler_old.dat'
rs = RuleScorer(scoring_class=RandomForestScorer(), scaling_class=MinMaxScaler(min_value=-100, max_value=0))
cProfile.run('rs.fit(X_rules, y)', sort='cumtime', filename=f'{filename}.dat')

In [ ]:
p = pstats.Stats(f'{filename}.dat')
p.sort_stats('cumtime').print_stats()

~204 secs is mostly due to Sklearn's rf fit method. Can't really improve that - use parameters to reduce time instead.